--- /dev/null
+++ b/src/main/java/me/earthme/luminol/commands/LuminolConfigCommand.java
@@ -1,0 +_,121 @@
+package me.earthme.luminol.commands;
+
+import me.earthme.luminol.config.LuminolConfig;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.TextColor;
+import org.bukkit.Location;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class LuminolConfigCommand extends Command {
+    public LuminolConfigCommand() {
+        super("luminolconfig");
+        this.setPermission("luminol.commands.luminolconfig");
+        this.setDescription("Manage config file");
+        this.setUsage("/luminolconfig");
+    }
+
+    public void wrongUse(CommandSender sender) {
+        sender.sendMessage(
+                Component
+                        .text("Wrong use!")
+                        .color(TextColor.color(255, 0, 0))
+        );
+    }
+
+    @Override
+    public @NotNull List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, @NotNull String[] args, @Nullable Location location) throws IllegalArgumentException {
+        final List<String> result = new ArrayList<>();
+
+        if (args.length == 1) {
+            result.add("query");
+            result.add("set");
+            result.add("reset");
+            result.add("reload");
+        } else if (args.length == 2) {
+            result.addAll(LuminolConfig.getAllConfigPaths(args[1]));
+        }
+        return result;
+    }
+
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        if (!this.testPermission(sender)) {
+            sender.sendMessage(Component
+                    .text("No permission to execute this command!")
+                    .color(TextColor.color(255, 0, 0))
+            );
+        }
+
+        if (args.length < 1) {
+            wrongUse(sender);
+            return true;
+        }
+
+        switch (args[0]) {
+            case "reload" -> {
+                LuminolConfig.reloadAsync().thenAccept(nullValue -> sender.sendMessage(
+                        Component
+                                .text("Reloaded config file!")
+                                .color(TextColor.color(0, 255, 0))
+                ));
+            }
+            case "set" -> {
+                if (args.length == 2 || args.length > 3) {
+                    wrongUse(sender);
+                    return true;
+                } else if (LuminolConfig.setConfig(args[1], args[2])) {
+                    LuminolConfig.reloadAsync().thenAccept(nullValue -> sender.sendMessage(
+                            Component
+                                    .text("Set Config " + args[1] + " to " + args[2] + " successfully!")
+                                    .color(TextColor.color(0, 255, 0))
+                    ));
+                } else {
+                    sender.sendMessage(
+                            Component
+                                    .text("Failed to set config " + args[1] + " to " + args[2] + "!")
+                                    .color(TextColor.color(255, 0, 0))
+                    );
+                }
+            }
+            case "reset" -> {
+                if (args.length != 2) {
+                    wrongUse(sender);
+                    return true;
+                } else {
+                    LuminolConfig.resetConfig(args[1]);
+                    LuminolConfig.reloadAsync().thenAccept(nullValue -> sender.sendMessage(
+                            Component
+                                    .text("Reset Config " + args[1] + " to " + LuminolConfig.getConfig(args[1]) + " successfully!")
+                                    .color(TextColor.color(0, 255, 0))
+                    ));
+                }
+            }
+            case "query" -> {
+                if (args.length != 2) {
+                    wrongUse(sender);
+                    return true;
+                } else {
+                    sender.sendMessage(
+                            Component
+                                    .text("Config " + args[1] + " is " + LuminolConfig.getConfig(args[1]) + "!")
+                                    .color(TextColor.color(0, 255, 0))
+                    );
+                }
+            }
+
+            default -> sender.sendMessage(
+                    Component
+                            .text("Unknown action!")
+                            .color(TextColor.color(255, 0, 0))
+            );
+        }
+
+        return true;
+    }
+}
