--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/modules/misc/CollisionBehaviorConfig.java
@@ -1,0 +_,25 @@
+package me.earthme.luminol.config.modules.misc;
+
+import me.earthme.luminol.config.EnumConfigCategory;
+import me.earthme.luminol.config.IConfigModule;
+import me.earthme.luminol.config.flags.ConfigInfo;
+
+public class CollisionBehaviorConfig implements IConfigModule {
+    @ConfigInfo(baseName = "mode", comments =
+            """
+                    Available Value:
+                    VANILLA
+                    BLOCK_SHAPE_VANILLA
+                    PAPER""")
+    public static String behaviorMode = "BLOCK_SHAPE_VANILLA";
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.MISC;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "collision_behavior";
+    }
+}
