--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/modules/misc/RaidChangesConfig.java
@@ -1,0 +_,54 @@
+package me.earthme.luminol.config.modules.misc;
+
+import me.earthme.luminol.config.ConfigInfo;
+import me.earthme.luminol.config.EnumConfigCategory;
+import me.earthme.luminol.config.IConfigModule;
+
+public class RaidChangesConfig implements IConfigModule {
+    @ConfigInfo(baseName = "allow-bad-omen-trigger-raid", comments =
+            """
+                    Allow players with ominous signs to skip a\s
+                    30-second cooldown and trigger attacks directly""")
+    public static boolean trigger = false;
+
+    @ConfigInfo(baseName = "give-bad-omen-when-kill-patrol-leader", comments =
+            """
+                    Enable players to obtain an ominous omen\s
+                    effect when killing the patrol team captain""")
+    public static boolean effect = false;
+
+    @ConfigInfo(baseName = "bad-omen-infinite", comments =
+            """
+                    Enable bad omen effect infinite time
+                        --- this config is not old version's function""")
+    public static boolean infinite = false;
+
+    @ConfigInfo(baseName = "skip-height-check", comments =
+            """
+                    Disable y <= 96 check.
+                    If you enabled revert_274911, this config will useless
+                    and always behavior of enabled""")
+    public static boolean height_check = false;
+
+    @ConfigInfo(baseName = "skip-self-raid-check", comments =
+            """
+                    Disable raid self check\s
+                        --- this config is not old version's function""")
+    public static boolean self_check = false;
+
+    @ConfigInfo(baseName = "revert-274911", comments =
+            """
+                    Revert Old raid's find spawn position logic
+                        --- This revert MC-274911""")
+    public static boolean pos_revert = false;
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.MISC;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "revert-raid-changes";
+    }
+}
