From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 29 Apr 2025 23:03:56 +0800
Subject: [PATCH] Cross Region Damage trace


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index d0853f44f1b4c88ba62a74db9e49a32709685425..ac8353216c10f8d87a7a25e1338aed686490f139 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1378,6 +1378,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.awardStat(Stats.ENTITY_KILLED_BY.get(killCredit.getType()));
             killCredit.awardKillScore(this, cause);
             this.createWitherRose(killCredit);
+        // Luminol Start - Cross Region Damage trace
+        } else {
+            final LivingEntity entitylivingnew = this.getKillCreditOrigin();
+            if (entitylivingnew != null) {
+                this.damageTransferToAsync(entitylivingnew, cause);
+            }
+        // Luminol End - Cross Region Damage trace
         }
 
         this.level().broadcastEntityEvent(this, (byte)3);
@@ -1392,6 +1399,20 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.setClientLoaded(false);
     }
 
+    // Luminol Start - Cross Region Damage trace
+    private void damageTransferToAsync(LivingEntity entity, DamageSource cause) {
+        entity.getBukkitEntity().taskScheduler.schedule((LivingEntity nmsEntity) -> {
+            try {
+                this.awardStat(Stats.ENTITY_KILLED_BY.get(nmsEntity.getType()));
+                nmsEntity.awardKillScore(this, cause);
+                this.createWitherRose(nmsEntity);
+            } catch (Throwable ex) {
+                LOGGER.error(ex.toString(), ex);
+            }
+        }, null, 1L );
+    }
+    // Luminol End - Cross Region Damage trace
+
     // Leaves start - exp fix
     private boolean shouldDropExperience(boolean eventResult, boolean forceUseEvent) {
         if (forceUseEvent) {
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 3e227c644072bfe9ec16078adc2b685edfee83c6..a9b3b3f594f11215060b370c338b1fedd97a6787 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -141,7 +141,6 @@ import org.slf4j.Logger;
 // CraftBukkit start
 import java.util.ArrayList;
 import java.util.HashSet;
-import java.util.Set;
 import java.util.LinkedList;
 import java.util.UUID;
 import net.minecraft.world.item.component.Consumable;
@@ -152,7 +151,6 @@ import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
-import org.bukkit.event.entity.EntityKnockbackEvent;
 import org.bukkit.event.entity.EntityPotionEffectEvent;
 import org.bukkit.event.entity.EntityRegainHealthEvent;
 import org.bukkit.event.entity.EntityRemoveEvent;
@@ -1853,6 +1851,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
             final LivingEntity entityliving = this.getKillCredit();
             if (entityliving != null) {
                 entityliving.awardKillScore(this, damageSource);
+            // Luminol Start - Cross Region Damage trace
+            } else {
+                final LivingEntity entitylivingnew = this.getKillCreditOrigin();
+                if (entitylivingnew != null) {
+                    this.damageTransferToAsync(entitylivingnew, damageSource);
+                }
+            // Luminol End - Cross Region Damage trace
             }
         }); // Paper end
         this.postDeathDropItems(deathEvent); // Paper
@@ -1863,6 +1868,18 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return deathEvent; // Paper
     }
 
+    // Luminol Start - Cross Region Damage trace
+    private void damageTransferToAsync(LivingEntity entity, DamageSource damageSource) {
+        entity.getBukkitEntity().taskScheduler.schedule((LivingEntity nmsEntity) -> {
+            try {
+                nmsEntity.awardKillScore(this, damageSource);
+            } catch (Throwable ex) {
+                LOGGER.error(ex.toString(), ex);
+            }
+        }, null, 1L );
+    }
+    // Luminol End - Cross Region Damage trace
+
     protected void dropEquipment(ServerLevel level) {
     }
     protected void postDeathDropItems(org.bukkit.event.entity.EntityDeathEvent event) {} // Paper - method for post death logic that cannot be ran before the event is potentially cancelled
@@ -2496,6 +2513,18 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
     }
 
+    // Luminol Start - Cross Region Damage trace
+    @Nullable
+    public LivingEntity getKillCreditOrigin() {
+        if (this.lastHurtByPlayer != null) {
+            return this.lastHurtByPlayer;
+        } else if (this.lastHurtByMob != null) {
+            return this.lastHurtByMob;
+        }
+        return null;
+    }
+    // Luminol End - Cross Region Damage trace
+
     public final float getMaxHealth() {
         return (float)this.getAttributeValue(Attributes.MAX_HEALTH);
     }
