From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 10 May 2025 19:46:16 +0800
Subject: [PATCH] Fix creative player item pick


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 38f0aacba93236b8e5c19bb729326feac3285c94..a9cc35b4b253d9a7fd90f811af9be629b164fad0 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -424,7 +424,15 @@ public class ItemEntity extends Entity implements TraceableEntity {
             Item item1 = item.getItem();
             int count = item.getCount();
             // CraftBukkit start - fire PlayerPickupItemEvent
-            int canHold = entity.getInventory().canHold(item);
+            // Luminol start - Fix creative item picking
+            int canHold;
+            if (entity instanceof net.minecraft.server.level.ServerPlayer
+                    && ((net.minecraft.server.level.ServerPlayer) entity).gameMode.getGameModeForPlayer().equals(net.minecraft.world.level.GameType.CREATIVE)) {
+                canHold = count;
+            } else {
+                canHold = entity.getInventory().canHold(item);
+            }
+            // Luminol end
             int remaining = count - canHold;
             boolean flyAtPlayer = false; // Paper
 
