From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 29 Apr 2025 18:55:06 +0800
Subject: [PATCH] EndGateway Portal Passengers Fix


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 920b7a3e6def4ac0078a1e543d7c17f5d4955fc5..51cad9bbf446d193f85a897c33115dfcd0305f57 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3829,20 +3829,27 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
             EntityTreeNode curr;
             while ((curr = queue.pollFirst()) != null) {
-                EntityTreeNode[] passengers = curr.passengers;
+                restore(curr); // Luminol - EndGateway Portal Passengers Fix
+            }
+        }
+
+        // Luminol Start - EndGateway Portal Passengers Fix
+        private EntityTreeNode[] restore(EntityTreeNode entity) {
+                EntityTreeNode[] passengers = entity.passengers;
                 if (passengers == null) {
-                    continue;
+                    return null;
                 }
 
                 List<Entity> newPassengers = new java.util.ArrayList<>();
                 for (EntityTreeNode passenger : passengers) {
+                    passenger.passengers = restore(passenger);
                     newPassengers.add(passenger.root);
-                    passenger.root.vehicle = curr.root;
+                    passenger.root.vehicle = entity.root;
                 }
-
-                curr.root.passengers = ImmutableList.copyOf(newPassengers);
-            }
+                entity.root.passengers = ImmutableList.copyOf(newPassengers);
+                return passengers;
         }
+        // Luminol End - EndGateway Portal Passengers Fix
 
         public void addTracker() {
             for (final EntityTreeNode node : this.getFullTree()) {
