From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Tue, 29 Apr 2025 23:03:56 +0800
Subject: [PATCH] Cross Region Damage trace


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index d0853f44f1b4c88ba62a74db9e49a32709685425..e1945ab589b216261b841f20bdf4a8a65d64bb82 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1378,6 +1378,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.awardStat(Stats.ENTITY_KILLED_BY.get(killCredit.getType()));
             killCredit.awardKillScore(this, cause);
             this.createWitherRose(killCredit);
+        // Luminol Start - Cross Region Damage trace
+        } else {
+            final LivingEntity entitylivingnew = this.getKillCreditOrigin();
+            if (entitylivingnew != null) {
+                this.damageTransferToAsync(entitylivingnew, cause);
+            }
+        // Luminol End - Cross Region Damage trace
         }
 
         this.level().broadcastEntityEvent(this, (byte)3);
@@ -1392,6 +1399,20 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.setClientLoaded(false);
     }
 
+    // Luminol Start - Cross Region Damage trace
+    private void damageTransferToAsync(LivingEntity entity, DamageSource cause) {
+        entity.getBukkitEntity().taskScheduler.schedule((LivingEntity nmsEntity) -> {
+            try {
+                this.awardStat(Stats.ENTITY_KILLED_BY.get(nmsEntity.getType()));
+                nmsEntity.awardKillScore(this, cause);
+                this.createWitherRose(nmsEntity);
+            } catch (Throwable ex) {
+                LOGGER.error(ex.getMessage(), ex);
+            }
+        }, null, 1L );
+    }
+    // Luminol End - Cross Region Damage trace
+
     // Leaves start - exp fix
     private boolean shouldDropExperience(boolean eventResult, boolean forceUseEvent) {
         if (forceUseEvent) {
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 58c28101b3d0bbb7d80b8c42a861712ddd1d10ec..3a5a2d7b91a1abb952bb6e1a72b77de3cd3dac39 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1853,6 +1853,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
             final LivingEntity entityliving = this.getKillCredit();
             if (entityliving != null) {
                 entityliving.awardKillScore(this, damageSource);
+            // Luminol Start - Cross Region Damage trace
+            } else {
+                final LivingEntity entitylivingnew = this.getKillCreditOrigin();
+                if (entitylivingnew != null) {
+                    this.damageTransferToAsync(entitylivingnew, damageSource);
+                }
+            // Luminol End - Cross Region Damage trace
             }
         }); // Paper end
         this.postDeathDropItems(deathEvent); // Paper
@@ -1863,6 +1870,18 @@ public abstract class LivingEntity extends Entity implements Attackable {
         return deathEvent; // Paper
     }
 
+    // Luminol Start - Cross Region Damage trace
+    private void damageTransferToAsync(LivingEntity entity, DamageSource damageSource) {
+        entity.getBukkitEntity().taskScheduler.schedule((LivingEntity nmsEntity) -> {
+            try {
+                nmsEntity.awardKillScore(this, damageSource);
+            } catch (Throwable ex) {
+                LOGGER.error(ex.getMessage(), ex);
+            }
+        }, null, 1L );
+    }
+    // Luminol End - Cross Region Damage trace
+
     protected void dropEquipment(ServerLevel level) {
     }
     protected void postDeathDropItems(org.bukkit.event.entity.EntityDeathEvent event) {} // Paper - method for post death logic that cannot be ran before the event is potentially cancelled
@@ -2496,6 +2515,18 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
     }
 
+    // Luminol Start - Cross Region Damage trace
+    @Nullable
+    public LivingEntity getKillCreditOrigin() {
+        if (this.lastHurtByPlayer != null) {
+            return this.lastHurtByPlayer;
+        } else if (this.lastHurtByMob != null) {
+            return this.lastHurtByMob;
+        }
+        return null;
+    }
+    // Luminol End - Cross Region Damage trace
+
     public final float getMaxHealth() {
         return (float)this.getAttributeValue(Attributes.MAX_HEALTH);
     }
